bser/tap.go

```diff
@@ -0,0 +1,68 @@
+package bser
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// Tap sends all PDUs sent in either direction along rw
+// through rfn for reads and wfn for writes
+func Tap(rw io.ReadWriter, rfn func([]byte), wfn func([]byte)) io.ReadWriter {
+	var r io.Reader = rw
+	var w io.Writer = rw
+
+	if rfn != nil {
+		r = io.TeeReader(rw, logWriter(rfn))
+	}
+	if wfn != nil {
+		w = io.MultiWriter(rw, logWriter(wfn))
+	}
+
+	return tap{r, w}
+}
+
+type tap struct {
+	io.Reader
+	io.Writer
+}
+
+func logWriter(fn func([]byte)) io.Writer {
+	pr, pw := io.Pipe()
+	go func() {
+		for {
+			buf, err := readPDU(pr)
```

Comment:
won't this continually wait unless we somehow call `pw.Close()` which will trigger an `io.EOF`  error?